src/rust/engine/src/scheduler.rs

```diff
@@ -295,64 +297,92 @@ impl Scheduler {
+    let results = match session.maybe_display() {
+      Some(display) => {
+        {
+          let mut display = display.lock();
+          display.start();
+        }
+        let unique_handle = LOGGER.register_engine_display(display.clone());
+
+        let results = loop {
+          if let Ok(res) = receiver.recv_timeout(Duration::from_millis(100)) {
+            break res;
+          } else {
+            Scheduler::display_ongoing_tasks(
+              &self.core.graph,
+              &roots,
+              display,
+              &mut tasks_to_display,
+            );
+          }
+        };
+        {
+          let mut display = display.lock();
+          display.finish();
+        }
+        LOGGER.deregister_engine_display(unique_handle);
+        results
       }
+      None => loop {
+        if let Ok(res) = receiver.recv_timeout(Duration::from_millis(100)) {
+          break res;
+        }
+      },
     };
-    if let Some(ref mut display) = maybe_display {
-      display.finish();
-    };
-
     results
   }
 
   fn display_ongoing_tasks(
     graph: &Graph<NodeKey>,
     roots: &[NodeKey],
-    display: &mut EngineDisplay,
+    display: &Mutex<EngineDisplay>,
     tasks_to_display: &mut IndexMap<String, Duration>,
   ) {
     // Update the graph. To do that, we iterate over heavy hitters.
-    let heavy_hitters = graph.heavy_hitters(&roots, display.worker_count());
+
+    let worker_count = {
+      let d = display.lock();
+      d.worker_count()
+    };
+    let heavy_hitters = graph.heavy_hitters(&roots, worker_count);
+
     // Insert every one in the set of tasks to display.
     // For tasks already here, the durations are overwritten.
     tasks_to_display.extend(heavy_hitters.clone().into_iter());
+
     // And remove the tasks that no longer should be there.
     for (task, _) in tasks_to_display.clone().into_iter() {
       if !heavy_hitters.contains_key(&task) {
         tasks_to_display.swap_remove(&task);
       }
     }
-    let display_worker_count = display.worker_count();
-    let ongoing_tasks = tasks_to_display;
-    for (i, id) in ongoing_tasks.iter().enumerate() {
+
+    for (i, id) in tasks_to_display.iter().enumerate() {
       // TODO Maybe we want to print something else besides the ID here.
-      display.update(i.to_string(), format!("{:?}", id));
+      let mut d = display.lock();
+      d.update(i.to_string(), format!("{:?}", id));
     }
+
     // If the number of ongoing tasks is less than the number of workers,
     // fill the rest of the workers with empty string.
     // TODO(yic): further improve the UI. https://github.com/pantsbuild/pants/issues/6666
-    for i in ongoing_tasks.len()..display_worker_count {
-      display.update(i.to_string(), "".to_string());
+    let worker_count = {
+      let d = display.lock();
+      d.worker_count()
+    };
+    for i in tasks_to_display.len()..worker_count {
```

Comment:
It should be ok to keep the lock acquired around this entire series of updates and the `render` call.